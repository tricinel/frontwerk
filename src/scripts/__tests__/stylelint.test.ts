/* eslint-disable global-require */
import cases from 'jest-in-case';
import jestSerializerPath from 'jest-serializer-path';

import {
  unquoteSerializer,
  winPathSerializer
} from '../../helpers/serializers';

expect.addSnapshotSerializer(unquoteSerializer);
expect.addSnapshotSerializer(winPathSerializer);
expect.addSnapshotSerializer(jestSerializerPath);

jest.mock('consola');

const testCases = [
  {
    name: 'calls prettier CLI with default args'
  },
  {
    name: '--no-cache will disable caching',
    args: ['--no-cache']
  },
  {
    name: '--no-color will disable colored output',
    args: ['--no-color']
  },
  {
    name: 'calls prettier CLI with args',
    args: ['my-src/**/*.css']
  },
  {
    name: 'does not use built-in config with .stylelintrc file',
    fileExists: (filename: string): boolean => filename === '.stylelintrc'
  },
  {
    name: 'does not use built-in config with stylelint.config.js file',
    fileExists: (filename: string): boolean =>
      filename === 'stylelint.config.js'
  },
  {
    name: 'does not use built-in config with stylelint pkg prop',
    hasPkgProp: (prop: string): boolean => prop === 'stylelint'
  },
  {
    name: 'does not use built-in ignore with --ignore-path',
    args: ['--ignore-path', './my-ignore']
  },
  {
    name: 'does not use built-in config with --config',
    args: ['--config', './custom-config.js']
  }
];

const testFn = ({
  fileExists = () => false,
  hasPkgProp = () => false,
  args = []
}) => {
  const crossSpawn = require('cross-spawn');
  const syncSpy = jest
    .spyOn(crossSpawn, 'sync')
    .mockImplementation(() => ({ status: 0 }));
  const exitSpy = jest.spyOn(process, 'exit');
  const { argv: originalArgv } = process;

  Object.assign(require('../../utils/fileExists'), { default: fileExists });
  Object.assign(require('../../utils/pkg'), { hasPkgProp });
  Object.assign(require('../../utils/resolveBin'), {
    resolveBin: (modName: string, { executable = modName } = {}) => executable
  });

  try {
    // Tests
    process.argv = ['node', '../stylelint', ...args];

    require('../stylelint');

    expect(syncSpy).toHaveBeenCalledTimes(1);
    const [firstCall] = syncSpy.mock.calls;
    const [script, calledArgs] = firstCall as [string, string[]];
    expect([script, ...calledArgs].join(' ')).toMatchSnapshot();
  } finally {
    // We reset everything afterEach
    exitSpy.mockRestore();
    syncSpy.mockRestore();
    process.argv = originalArgv;
    jest.resetModules();
  }
};

cases('stylelint', testFn, testCases);
