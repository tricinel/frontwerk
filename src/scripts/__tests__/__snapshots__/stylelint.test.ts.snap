// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stylelint --no-cache will disable caching 1`] = `stylelint **/*.+(css|sass|scss) --config ./src/config/stylelint.config.js --ignore-path ./src/config/stylelintignore --color --no-cache`;

exports[`stylelint --no-color will disable colored output 1`] = `stylelint **/*.+(css|sass|scss) --config ./src/config/stylelint.config.js --ignore-path ./src/config/stylelintignore --cache --no-color`;

exports[`stylelint calls prettier CLI with args 1`] = `stylelint --config ./src/config/stylelint.config.js --ignore-path ./src/config/stylelintignore --cache --color my-src/**/*.css`;

exports[`stylelint calls prettier CLI with default args 1`] = `stylelint **/*.+(css|sass|scss) --config ./src/config/stylelint.config.js --ignore-path ./src/config/stylelintignore --cache --color`;

exports[`stylelint does not use built-in config with .stylelintrc file 1`] = `stylelint **/*.+(css|sass|scss) --ignore-path ./src/config/stylelintignore --cache --color`;

exports[`stylelint does not use built-in config with --config 1`] = `stylelint **/*.+(css|sass|scss) --ignore-path ./src/config/stylelintignore --cache --color --config ./custom-config.js`;

exports[`stylelint does not use built-in config with stylelint pkg prop 1`] = `stylelint **/*.+(css|sass|scss) --ignore-path ./src/config/stylelintignore --cache --color`;

exports[`stylelint does not use built-in config with stylelint.config.js file 1`] = `stylelint **/*.+(css|sass|scss) --ignore-path ./src/config/stylelintignore --cache --color`;

exports[`stylelint does not use built-in ignore with --ignore-path 1`] = `stylelint **/*.+(css|sass|scss) --config ./src/config/stylelint.config.js --cache --color --ignore-path ./my-ignore`;
